File: internal/bridges/cache_bridge.go
package bridges

import (
	"sync"

	"github.com/dop251/goja"
)

type CacheBridge struct {
	name string
	mu   sync.RWMutex
	data map[string]any
}

func NewCacheBridge(name string) *CacheBridge {
	return &CacheBridge{
		name: name,
		data: make(map[string]any),
	}
}

func (cb *CacheBridge) Register(vm *goja.Runtime, hostObj *goja.Object) error {
	cacheObj := vm.NewObject()

	_ = cacheObj.Set("get", func(key string) any {
		cb.mu.RLock()
		defer cb.mu.RUnlock()
		return cb.data[key]
	})

	_ = cacheObj.Set("set", func(key string, val any) {
		cb.mu.Lock()
		defer cb.mu.Unlock()
		cb.data[key] = val
	})

	_ = cacheObj.Set("delete", func(key string) {
		cb.mu.Lock()
		defer cb.mu.Unlock()
		delete(cb.data, key)
	})

	_ = cacheObj.Set("clear", func() {
		cb.mu.Lock()
		defer cb.mu.Unlock()
		for k := range cb.data {
			delete(cb.data, k)
		}
	})

	return hostObj.Set("cache", cacheObj)
}

File: YStreamUtils/Core/Bridges/CacheBridge.cs
using Jint;
using Jint.Native;
using Jint.Native.Json;
using YStreamUtils.Core.Data;
using YStreamUtils.Core.Entities;

namespace YStreamUtils.Core.Bridges;

public class CacheBridge(string name, AppDbContext dbContext) : IBridge
{
    private readonly ReaderWriterLockSlim _lock = new();

    private string CreateKey(string key) => $"{name}:{key}";
    
    public void Register(Engine vm, JsObject hostObj)
    {
        var cacheObj = new JsObject(vm);

        cacheObj.Set("get", JsValue.FromObject(vm, new Func<string, JsValue>((key) =>
        {
            _lock.EnterReadLock();
            try
            {
                var fullKey = CreateKey(key);
                var cacheEntry = dbContext.Caches.FirstOrDefault(x => x.Key == fullKey);
                
                if (cacheEntry == null || string.IsNullOrEmpty(cacheEntry.Value))
                {
                    return JsValue.Null;
                }

                var parser = new JsonParser(vm);
                return parser.Parse(cacheEntry.Value);
            }
            finally
            {
                _lock.ExitReadLock();
            }
        })));

        cacheObj.Set("set", JsValue.FromObject(vm, new Action<string, JsValue>((key, val) =>
        {
            _lock.EnterWriteLock();
            try
            {
                var fullKey = CreateKey(key);
                
                var serializer = new JsonSerializer(vm);
                JsValue stringified = serializer.Serialize(val, JsValue.Undefined, JsValue.Undefined);
                string stringValue = stringified.AsString();

                var cacheEntry = dbContext.Caches.FirstOrDefault(x => x.Key == fullKey);
                if (cacheEntry != null)
                {
                    cacheEntry.Value = stringValue;
                }
                else
                {
                    dbContext.Caches.Add(new Cache 
                    { 
                        Key = fullKey, 
                        Value = stringValue 
                    });
                }
                
                dbContext.SaveChanges();
            }
            finally
            {
                _lock.ExitWriteLock();
            }
        })));

        cacheObj.Set("delete", JsValue.FromObject(vm, new Action<string>((key) =>
        {
            _lock.EnterWriteLock();
            try
            {
                var fullKey = CreateKey(key);
                var cacheEntry = dbContext.Caches.FirstOrDefault(x => x.Key == fullKey);
                
                if (cacheEntry != null)
                {
                    dbContext.Caches.Remove(cacheEntry);
                    dbContext.SaveChanges();
                }
            }
            finally
            {
                _lock.ExitWriteLock();
            }
        })));

        cacheObj.Set("clear", JsValue.FromObject(vm, new Action(() =>
        {
            _lock.EnterWriteLock();
            try
            {
                var prefix = $"{name}:";
                var entriesToRemove = dbContext.Caches.Where(x => x.Key.StartsWith(prefix));
                
                dbContext.Caches.RemoveRange(entriesToRemove);
                dbContext.SaveChanges();
            }
            finally
            {
                _lock.ExitWriteLock();
            }
        })));

        hostObj.Set("cache", cacheObj);
    }
}
